File: workers/ts/src/protocol.ts
export type File = { path: string; content: string };
export type Snapshot = { files: File[]; project?: string | null };

export type Op = {
  id: string;
  schemaVersion: 1;
  type: string;
  target: { symbolId: string; addressId?: string | null };
  params: any;
  guards: any;
  effects: any;
  provenance: { rev?: string; author?: string; timestamp?: string };
};

export type BuildAndDiffParams = {
  base: Snapshot;
  left: Snapshot;
  right: Snapshot;
  config: { deterministicSeed?: string };
};

export type BuildAndDiffResult = {
  opLogLeft: Op[];
  opLogRight: Op[];
  symbolMaps: Record<string, Array<{ symbolId: string; addressId: string }>>;
  diagnostics: any[];
};

File: semmerge/lang/ts/protocol.py
"""Typed structures for communicating with the TypeScript worker."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, List, Optional


@dataclass
class ProgramSnapshot:
    files: List[Dict[str, str]]
    project: Optional[str] = None


@dataclass
class BuildAndDiffResult:
    opLogLeft: List[Dict[str, Any]]
    opLogRight: List[Dict[str, Any]]
    symbolMaps: Dict[str, Any]
    diagnostics: List[Dict[str, Any]]
